commit 7f8a212698d49413af6f2682585a6bd7a3279fc4
Author: Svante Signell <svante.signell@gmail.com>
Date:   Sat Jun 27 00:30:56 2020 +0200

    pflocal: Add support for getsockopt(SO_{RECV,SND}BUF)
    
    * S_socket_getopt: Handle SO_RCVBUF and SO_SNDBUF cases.

commit b95271e96190f0b854b46c8d5b8aab1a812d35a8
Author: Samuel Thibault <samuel.thibault@ens-lyon.org>
Date:   Sat Jun 27 00:37:01 2020 +0200

    pflocal: Fix crash on passing erroneous port
    
    * pflocal/socket.c (S_socket_getopt): Read `user->sock' after checking
      `user'.

commit b290be35e0717ad68d9a851974e7fcfedc88c448
Author: Samuel Thibault <samuel.thibault@ens-lyon.org>
Date:   Sat Jun 27 01:12:14 2020 +0200

    pflocal: Add support for setsockopt(SO_{RECV,SND}BUF)
    
    Thanks Svante Signell for the initial patch.
    
    * libpipe/pipe.c (pipe_recv): Move writer wake code to...
    (_pipe_wake_writers): ... new function.
    * libpipe/pipe.h (_pipe_wake_writers): New prototype.
    * pflocal/sock.h (PFLOCAL_WRITE_LIMIT_MAX): New macro.
    * pflocal/socket.c (S_socket_setopt): Handle SO_RCVBUF and SO_SNDBUF cases.

commit 414f6c35a62a7e72f40c39089a9952913a7eeb90
Author: Samuel Thibault <samuel.thibault@ens-lyon.org>
Date:   Sun Jul 5 02:32:44 2020 +0200

    pflocal: return ENOTCONN on get/setpot(SO_SNDBUF) on unconnected socket
    
    EPIPE would raise SIGPIPE, which applications do not expect to happen on
    a mere get/setsockopt.
    
    * pflocal/socket.c (S_socket_getopt, S_socket_setopt): Return ENOTCONN
    instead of EPIPE.

commit 9b64495c6cd22cd65f118c18a36b2de7cbde4344
Author: Samuel Thibault <samuel.thibault@ens-lyon.org>
Date:   Sun Jul 5 17:09:58 2020 +0200

    pflocal: support SO_SNDBUF for unconnected sockets
    
    We can store the requested value, to be applied when we connect the
    sockets.
    
    * pflocal/sock.h (struct sock): Add req_write_limit field.
    * pflocal/sock.c (sock_create): Initialize req_write_limit field to 0.
    (sock_connect): Bump the write_limit of the write pipe to the
    req_write_limit value.
    (sock_shutdown): Update req_write_limit from the write_limit of the
    write pipe.
    * pflocal/socket.c (S_socket_getopt, S_socket_setopt): When write_pipe
    is NULL, use req_write_limit.

Index: hurd-debian/libpipe/pipe.c
===================================================================
--- hurd-debian.orig/libpipe/pipe.c
+++ hurd-debian/libpipe/pipe.c
@@ -200,6 +200,24 @@ void _pipe_no_writers (struct pipe *pipe
       pthread_mutex_unlock (&pipe->lock);
     }
 }
+
+/* Take any actions necessary when PIPE's writer can proceed.
+   PIPE should be locked. */
+void _pipe_wake_writers (struct pipe *pipe)
+{
+  pthread_cond_broadcast (&pipe->pending_writes);
+  pthread_mutex_unlock (&pipe->lock);
+
+  pthread_mutex_lock (&pipe->lock);	/* Get back the lock on PIPE.  */
+  /* Only wakeup selects if there's still writing space available.  */
+  if (pipe_readable (pipe, 1) < pipe->write_limit)
+    {
+      pthread_cond_broadcast (&pipe->pending_write_selects);
+      pipe_select_cond_broadcast (pipe);
+      /* We leave PIPE locked here, assuming the caller will soon unlock
+	 it and allow others access.  */
+    }
+}
 
 /* Return when either RPIPE is available for reading (if SELECT_READ is set
    in *SELECT_TYPE), or WPIPE is available for writing (if select_write is
@@ -474,18 +492,7 @@ pipe_recv (struct pipe *pipe, int nobloc
       timestamp (&pipe->read_time);
 
       /* And wakeup anyone that might be interested in it.  */
-      pthread_cond_broadcast (&pipe->pending_writes);
-      pthread_mutex_unlock (&pipe->lock);
-
-      pthread_mutex_lock (&pipe->lock);	/* Get back the lock on PIPE.  */
-      /* Only wakeup selects if there's still writing space available.  */
-      if (pipe_readable (pipe, 1) < pipe->write_limit)
-	{
-	  pthread_cond_broadcast (&pipe->pending_write_selects);
-	  pipe_select_cond_broadcast (pipe);
-	  /* We leave PIPE locked here, assuming the caller will soon unlock
-	     it and allow others access.  */
-	}
+      _pipe_wake_writers (pipe);
     }
 
   return err;
Index: hurd-debian/libpipe/pipe.h
===================================================================
--- hurd-debian.orig/libpipe/pipe.h
+++ hurd-debian/libpipe/pipe.h
@@ -263,6 +263,10 @@ void _pipe_no_readers (struct pipe *pipe
    should be locked.  */
 void _pipe_no_writers (struct pipe *pipe);
 
+/* Take any actions necessary when PIPE's writer can proceed.
+   PIPE should be locked. */
+void _pipe_wake_writers (struct pipe *pipe);
+
 extern void pipe_acquire_reader (struct pipe *pipe);
 
 extern void pipe_acquire_writer (struct pipe *pipe);
Index: hurd-debian/pflocal/sock.h
===================================================================
--- hurd-debian.orig/pflocal/sock.h
+++ hurd-debian/pflocal/sock.h
@@ -52,6 +52,9 @@ struct sock
      another socket.  */
   struct pipe *read_pipe, *write_pipe;
 
+  /* The write limit that this side would like write_pipe to support.  */
+  size_t req_write_limit;
+
   /* FLAGS from SOCK_*, below.  */
   unsigned flags;
 
@@ -187,4 +190,7 @@ error_t sock_global_shutdown ();
 extern struct port_class *sock_user_port_class;
 extern struct port_class *addr_port_class;
 
+/* Maximum allowed size for libpipe buffers */
+#define PFLOCAL_WRITE_LIMIT_MAX (1024*1024)
+
 #endif /* __SOCK_H__ */
Index: hurd-debian/pflocal/socket.c
===================================================================
--- hurd-debian.orig/pflocal/socket.c
+++ hurd-debian/pflocal/socket.c
@@ -434,11 +434,15 @@ S_socket_getopt (struct sock_user *user,
 		 data_t *value, size_t *value_len)
 {
   int ret = 0;
+  struct pipe *pipe;
+  struct sock *sock;
 
   if (!user)
     return EOPNOTSUPP;
 
-  pthread_mutex_lock (&user->sock->lock);
+  sock = user->sock;
+
+  pthread_mutex_lock (&sock->lock);
   switch (level)
     {
     case SOL_SOCKET:
@@ -450,7 +454,35 @@ S_socket_getopt (struct sock_user *user,
 	      ret = EINVAL;
 	      break;
 	    }
-	  *(int *)*value = user->sock->pipe_class->sock_type;
+	  *(int *)*value = sock->pipe_class->sock_type;
+	  *value_len = sizeof (int);
+	  break;
+	case SO_RCVBUF:
+	  if (*value_len < sizeof (int))
+	    {
+	      ret = EINVAL;
+	      break;
+	    }
+	  pipe = sock->read_pipe;
+	  if (!pipe)
+	    {
+	      ret = ENOTCONN;
+	      break;
+	    }
+	  *(int *)*value = pipe->write_limit;
+	  *value_len = sizeof (int);
+	  break;
+	case SO_SNDBUF:
+	  if (*value_len < sizeof (int))
+	    {
+	      ret = EINVAL;
+	      break;
+	    }
+	  pipe = sock->write_pipe;
+	  if (pipe)
+	    *(int *)*value = pipe->write_limit;
+	  else
+	    *(int *)*value = sock->req_write_limit;
 	  *value_len = sizeof (int);
 	  break;
 	case SO_ERROR:
@@ -481,7 +513,7 @@ S_socket_getopt (struct sock_user *user,
       ret = ENOPROTOOPT;
       break;
     }
-  pthread_mutex_unlock (&user->sock->lock);
+  pthread_mutex_unlock (&sock->lock);
 
   return ret;
 }
@@ -491,18 +523,98 @@ S_socket_setopt (struct sock_user *user,
 		 int level, int opt, data_t value, size_t value_len)
 {
   int ret = 0;
+  struct pipe *pipe;
+  struct sock *sock;
 
   if (!user)
     return EOPNOTSUPP;
 
-  pthread_mutex_lock (&user->sock->lock);
+  sock = user->sock;
+
+  pthread_mutex_lock (&sock->lock);
   switch (level)
     {
+    case SOL_SOCKET:
+      switch (opt)
+	{
+	case SO_RCVBUF:
+	  {
+	    int new, old;
+
+	    if (value_len < sizeof (int))
+	      {
+		ret = EINVAL;
+		break;
+	      }
+	    new = *(int *)value;
+	    if (new <= 0)
+	      {
+		ret = EINVAL;
+		break;
+	      }
+	    if (new > PFLOCAL_WRITE_LIMIT_MAX)
+	      new = PFLOCAL_WRITE_LIMIT_MAX;
+
+	    pipe = sock->read_pipe;
+	    if (!pipe)
+	      {
+		ret = ENOTCONN;
+		break;
+	      }
+
+	    pthread_mutex_lock (&pipe->lock);
+	    old = pipe->write_limit;
+	    pipe->write_limit = new;
+	    if (new > old)
+	      _pipe_wake_writers (pipe);
+	    pthread_mutex_unlock (&pipe->lock);
+	    break;
+	  }
+
+	case SO_SNDBUF:
+	  {
+	    int new, old;
+
+	    if (value_len < sizeof (int))
+	      {
+		ret = EINVAL;
+		break;
+	      }
+	    new = *(int *)value;
+	    if (new <= 0)
+	      {
+		ret = EINVAL;
+		break;
+	      }
+	    if (new > PFLOCAL_WRITE_LIMIT_MAX)
+	      new = PFLOCAL_WRITE_LIMIT_MAX;
+
+	    pipe = sock->write_pipe;
+	    if (!pipe)
+	      {
+		sock->req_write_limit = new;
+		break;
+	      }
+
+	    pthread_mutex_lock (&pipe->lock);
+	    old = pipe->write_limit;
+	    pipe->write_limit = new;
+	    if (new > old)
+	      _pipe_wake_writers (pipe);
+	    pthread_mutex_unlock (&pipe->lock);
+	    break;
+	  }
+
+	default:
+	  ret = ENOPROTOOPT;
+	  break;
+	}
+      break;
     default:
       ret = ENOPROTOOPT;
       break;
     }
-  pthread_mutex_unlock (&user->sock->lock);
+  pthread_mutex_unlock (&sock->lock);
 
   return ret;
 }
Index: hurd-debian/pflocal/sock.c
===================================================================
--- hurd-debian.orig/pflocal/sock.c
+++ hurd-debian/pflocal/sock.c
@@ -116,6 +116,7 @@ sock_create (struct pipe_class *pipe_cla
   new->refs = 0;
   new->flags = 0;
   new->write_pipe = NULL;
+  new->req_write_limit = 0;
   new->mode = mode;
   new->id = MACH_PORT_NULL;
   new->listen_queue = NULL;
@@ -397,6 +398,8 @@ sock_connect (struct sock *sock1, struct
 	  assert_backtrace (pipe);	/* Since PFLOCAL_SOCK_SHUTDOWN_READ isn't set.  */
 	  pipe_add_writer (pipe);
 	  wr->write_pipe = pipe;
+	  if (pipe->write_limit < wr->req_write_limit)
+	    pipe->write_limit = wr->req_write_limit;
 	}
     }
 
@@ -474,6 +477,8 @@ sock_shutdown (struct sock *sock, unsign
       /* Shutdown the write half.  */
       write_pipe = sock->write_pipe;
       sock->write_pipe = NULL;
+      if (write_pipe)
+	sock->req_write_limit = write_pipe->write_limit;
     }
 
   /* Unlock SOCK here, as we may subsequently wake up other threads. */
